chainermn/datasets/shuffle_datablock.py

```diff
@@ -0,0 +1,219 @@
+import itertools
+import random
+
+import chainer
+import functools
+import numpy
+
+
+def _increment_send_counts(table, send_rank, recv_rank, n=1):
+    key = (send_rank, recv_rank)
+    table.setdefault(key, 0)
+    table[key] += n
+
+
+def _flatten1(ary):
+    """1-level flatten"""
+    return functools.reduce(lambda a, b: a + b, ary)
+
+
+def _calc_new_length(data_lengths, block_lengths):
+    # the total number of elements over all processes, including
+    # already-scattered data.
+    # data_lengths: lengths of already-loaded data of all processes
+    # block_lengths: lengths of newly-loaded data blocks
+    size = len(data_lengths)
+    total = sum(data_lengths) + sum(block_lengths)
+
+    # new_length: new number of elements of all processes after shuffling
+    if total % size == 0:
+        new_lengths = [total // size] * size
+    else:
+        rem = total % size
+        rems = [1 if i < rem else 0 for i in range(size)]
+        new_lengths = [total // size + rems[i] for i in range(size)]
+
+    return new_lengths
+
+
+def _calc_alltoall_send_counts(data_lengths, block_lengths):
+    """Calculate send counts table from current data and loaded block length.
+
+    For example, if length of loaded blocks of all processes are
+       [1, 10, 18]
+    then we will perform MPI_alltoallv() communication so that lengths are
+       [10, 10, 9]   (force_equal_length is False)
+    or
+       [10, 10, 10]  (force_equal_length is True)
+    because total number of elements are 29.
+
+    In this case, if `force_equal_length` is False,
+    the communication matrix, namely send_counts, is like:
+       2 --> 0  (9 elements)
+    so
+       send_counts = {(2, 0): 9}
+
+    :param data_lengths: Already-loaded data length
+    :param block_lengths: Length of the newly loaded block
+    :param force_equal_length: If force equal length
+    :return: A dict of send counts
+    """
+    size = len(data_lengths)
+
+    new_lengths = _calc_new_length(data_lengths, block_lengths)
+
+    # If diff_length[rank] is >0, then the rank has more elements than
+    # expected, so it distribute the extra data to other ranks.
+    diff_length = [block_lengths[i] + data_lengths[i] - new_lengths[i]
+                   for i in range(size)]
+
+    send_counts = {}  # send_count table
+
+    # Need to calculate the number of elements sent <self->self>
+    # for MPI_Alltoallv().
```

Comment:
Doesn't self-to-self sending in mpi4py involve data copying or pickling? My worry is that this can be unnecessary performance loss or not.